### DISCUSSION TITLE
Enphase Microinverter Data Access & Usage Example

### DISCUSSION THREAD

**Discussion by SCHibbard** [jasonacox/Powerwall-Dashboard#467] [Show and tell]
More show & tell. I have Enphase inverters under each of my solar panels, and through their API I am able to access their output individually.  The inverters can be accessed from the local network, however a token is required that is good for a year. Access worked fine adding the info to Powerwall-Dashboard’s telegraf.local. Using the data I was able to make the panels below.  Notes:
1)	The API returns inverter serial numbers, so a mapping is needed from the installation that maps serial numbers to physical panel locations.  While the Enphase user account will show the array with data provided by the installer, Enphase themselves do not disclose serial number info to a customer.  I got the serial number mapping from my installer.
2)	The two array panels use Grafana’s Canvas Plug-In.  First general release of Canvas was Grafana 10 (I’m using 10.4.1).
3)	Replicating this for other installations will not be without effort currently.  I have not tried to figure out a way to build the canvas panels easily for other configurations.  Canvas’ editor has improved with each release, but it’s still a lot of manual work.

The project turned out well.  The Inverter Output Time Series Panel can get quite busy, so I mainly monitor the Canvas panels, however there are some interesting tidbits to be had from the time-series depiction.  For example, today, with a fairly clear sky, the effect of my neighbor’s chimney as the sun moves behind it can be seen on the lower panels outputs.  I was also glad I had it after a recent hail storm here.  Could easily see later if any damage occurred.

![Solar Panel Current Output Panel Image](https://github.com/jasonacox/Powerwall-Dashboard/assets/40345296/d13c1d10-249a-4687-9b4d-6ac2ab3da5ae)

![Solar Panel Production Panel Image](https://github.com/jasonacox/Powerwall-Dashboard/assets/40345296/86a74eee-4623-4ec4-8ec5-6d356f3ffc33)

![Solar Panel Time Series Panel Image](https://github.com/jasonacox/Powerwall-Dashboard/assets/40345296/ca293312-9bac-4c1c-8dc7-5255e3d9edd9)

**Comment by dam1n10n** [comment]
Hey SCHibbard, this is great! Would you be willing to share how you did this so that I can do it too? I have Enphase micro inverters and would like to see this data from my install. I am reasonably new to stuff like this and have been dipping my toes in with a mix of guides and discussions on here, google-fu and chatGPT. 

And by 'dipping my toes in' I mean I managed to add a new query to show a bar in the savings panel that shows my running electricity cost from my provider for the time period selected, I called it 'exported - imported'. Nothing fancy obviously, but you gotta start somewhere.

SELECT tg - fg FROM (SELECT sum("to_grid") * $avg_sell_per_kwh AS tg, sum("from_grid") * $avg_buy_per_kwh AS fg FROM kwh.http WHERE  $timeFilter tz('$tz'))

![image](https://github.com/jasonacox/Powerwall-Dashboard/assets/22421728/be8a5702-3dff-4ebe-ae49-839571046325)

**Reply by SCHibbard** [reply]
Hi!  Getting the data is straight forward, and the time-series panel (bottom one) is fairly easy to setup.  As I said above, the canvas panels (top two) are a bear.  I don't see a way to make that more universal, as each installation is graphically different.  Let me try to write something up for the data fetching and time-series.

**Reply by SCHibbard** [reply]
@dam1n10n: Want to be a guinea pig? I tried something.  See https://github.com/SCHibbard/enphase-dashboard.  It should create dashboard JSONs very similar to what I showed above for inclusion in your main dashboard(s).  My coding is poor (just started playing with Python), so I think I will keep it separate from the Powerwall-Dashboard repo until the folks here think it is mature enough for inclusion. Let me know if it works!

**Reply by dam1n10n** [reply]
@SCHibbard, apologies for such a late reply. I am keen to give it a go, however I lack consistent time to be able to learn and try things. I will do my best to be a guinea pig but I'm afraid I will screw up my powerwall dashboard as it currently is (not that I wouldn't be able to get it back, it would just be very frustrating). I very quickly get led down rabbit holes trying to learn linux etc, and it's almost like every time I get a chance to do something I have to find resources and re-learn stuff I've forgotten. I'm also very grateful for any instructions included in repositories here.  Ah well, I'll let you know how I go. Cheers

**Reply by khorton** [reply]
I installed your Enphase-Dashboard - Thank you very much for all your work to create and share this.  I'm struggling with one puzzling display issue - the current output and production for panel #30 (the last one) are not displayed.  If I inspect the data, I can see that grafana has credible data for both current output and production for that panel.  But, nothing is displayed.  It is almost like the text is the same color as the background, or something crazy like that.  I've studied the json, but cannot spot the error.  Any thoughts would be greatly appreciated.
<img width="246" alt="missing30" src="https://github.com/jasonacox/Powerwall-Dashboard/assets/77668/d330741d-ae68-489d-a316-f61562191a85">

**Reply by SCHibbard** [reply]
Hmm. I'll have to look at the JSON creation to see what the problem is. I use a pallet of colors to display the border of the box and the time series line (so they match). If I max out the pallet, I start over.  I have 27 panels, but I tested the pallet rollover by using a smaller number and it worked ok. I am in Croatia right now on vacation for a month, but heading back stateside next week. I'll take a look then. Sorry for delay, but thanks for testing/trying!

**Reply by SCHibbard** [reply]
Ahh, first cup of coffee down, thinking now. The color pallet looks to be working.  2 questions:
1) Is panel 30 showing on the time series panel (bottom one)?
2) Could you email me the dashboard JSON the program created? Note only panel serial numbers are in JSON, your Endphase login credentials are stored in telegraf.local, not the JSON. Email is schibbard@outlook.com. Maybe I can spot something from my phone!

**Reply by khorton** [reply]
Panel 30 is showing in the time series panel - both the legend and the line are visible.  I'll send the JSON to you.  But, you are on vacation.  This isn't that important.  It can wait a month.

**Reply by SCHibbard** [reply]
Body's on vacation, but brain multitasks. The values and panel numbers are aliased, then colored black or white to contrast against the background color, which is based on the wattage value. I'm guessing that's what's messed up. Can't remember from here if I tested with higher panel quantities after making that contrast change. Won't take a month, few days to catch up on mail & bills and I'll be back in the saddle!

**Reply by SCHibbard** [reply]
For any monitoring this, JSON creation bug found, fix verified by Kevin. Will update repository next week.

**Comment by khorton** [comment]
It is definitely an issue with 30Color.  It must be the same as the background, or transparent.  I manually edited the dashboard to have panel 30 use 29Color (the neighboring panel, so normally the output should be similar).  Now I can see the values and legends.  So no panic to fix this.

I also had to manually edit the placement of the "Total Current Output" and "Total Production" legends and values.  They ended up vertically centered in the midst of the panels, so unreadable.  I moved them to the top.  My Canvas grid is 6 high by 15 wide, if that matters.

Enjoy your vacation,

Kevin

**Reply by SCHibbard** [reply]
I'll take care of that too, probably left the vertical position fixed (oops)

**Comment by khorton** [comment]
Enphase data update frequency - I see that the data is updated every 5m (default defined in telegraf.enphase, but copied to telegraf.local during the setup process).  Are there any likely issues if this data was updated more frequently?  How long should it take to poll each microinverter?  Are they polled in parallel, or sequentially?

**Reply by SCHibbard** [reply]
Going from memory. I think Enohase docs say they update every 10 or 15 minutes. I set it faster than their updates. It's one call that brings back all connected inverters, not a separate call to each.

**Reply by SCHibbard** [reply]
Setting update faster will not get more data, as influx purges duplicate timestamps.

**Reply by khorton** [reply]
Thanks - I'll leave it alone

**Reply by khorton** [reply]
I played around with updates every 5s to learn how often the microinverters provide fresh data.  I learned that each microinverter seems to provide updated data approximately every 5m.  But, each microinverter updates at a different moment, so if you poll all microinverters, then poll again 10s later, you might find that one or more microinverters have new data.  In practice, there is probably no point to polling more often than every 5m.

Also, the total production from each microinverter updates at a different moment than the current output.
